### PR TITLE
Improve failure handling in the dynamic client

### LINKED ISSUE #359: Make request exponential backoff more predictable
Since #348 request retry logic is specified in multiple places. This can make retry logic much less predictable -- it is vastly easier to predict behavior when retry logic is specified in one place.

Before the next release, we should try to consolidate this into one place. Context: https://github.com/pulumi/pulumi-kubernetes/pull/348#discussion_r249188416

### PR DESCRIPTION
**Pull request by lblackstone**
- Rather than failing on partial resource discovery, process the partial list and
handle further failures downstream

Fixes #359

### CODE REVIEW

File: pkg/clients/clients.go [hunk 1]
```diff
@@ -118,23 +117,18 @@ func NewDynamicClientSet(clientConfig *rest.Config) (*DynamicClientSet, error) {
...
-		// TODO(lblackstone): Ensure that retry behavior is predictable.
-		err := retry.SleepingRetry(
-			func(i uint) (err error) {
-				m, err = dcs.RESTMapper.RESTMapping(gvk.GroupKind(), gvk.Version)
-				return
-			}).
-			WithMaxRetries(5).
-			WithBackoffFactor(2).
-			Do(meta.IsNoMatchError)
+		time.Sleep(2 * time.Second)
```
Comment (hausdorff) [inline pkg/clients/clients.go]: Do we actually need to sleep and retry here at all? It _seems_ like the invalidation and reset occur (a) synchronously, and (b) immediately?

File: pkg/clients/clients.go [hunk 2]
```diff
@@ -217,35 +216,54 @@ func (dcs *DynamicClientSet) namespaced(gvk schema.GroupVersionKind) (bool, erro
...
-				var resource *v1.APIResourceList
-				for _, gv := range gvs {
-					resource, err = dcs.DiscoveryClientCached.ServerResourcesForGroupVersion(gv.String())
-					if err != nil {
-						return
-					}
-
-					resources = append(resources, resource)
+func (dcs *DynamicClientSet) getServerResources(gvs ...schema.GroupVersion,
+) (resourceLists []*v1.APIResourceList, err error) {
```
Comment (hausdorff) [inline pkg/clients/clients.go]: I think that I will shoot myself in the foot with a "return an `error` _and_ a valid return value" strategy here. What do you think of passing a callback that lets you process each entry, as `os.Walk` does? Then it could be up to the caller to keep track of errors.

File: pkg/clients/clients.go (hunk 1 above)
Comment (lblackstone) [reply pkg/clients/clients.go]: Ah, you're right! I hadn't looked closely at the `Invalidate` method before. While digging, I also discovered that the `Reset` command calls `Invalidate`, so I'll remove the redundancy there.

File: pkg/clients/clients.go (hunk 2 above)
Comment (lblackstone) [reply pkg/clients/clients.go]: My thought was that the caller should be able to ignore GroupDiscoveryErrors, so it would only return an error in case of catastrophic failure. I can explicitly clear the resourceLists in those cases if you think that would be clearer.

Comment (hausdorff) [reply pkg/clients/clients.go]: I am not worried about that case so much as _any_ case where we return an `error` _and_ a return value. I would personally greatly prefer the `os.Walk` approach, which AFAIK is "the" idiom to use in these situations. I 100% promise you that if we do what's currently shown here, I will mess this up if I have to add a callsite for this function. :)

Comment (lblackstone) [reply pkg/clients/clients.go]: I took another pass and simplified the logic quite a bit. None of the functions return both valid data and an error at this point. LMK what you think.

File: pkg/clients/clients.go [hunk 3]
```diff
@@ -162,102 +165,45 @@ func (dcs *DynamicClientSet) gvkForKind(kind Kind) (gvk schema.GroupVersionKind,
...
+	resourceList, err := dcs.DiscoveryClientCached.ServerResourcesForGroupVersion(gv.String())
 
 	if err != nil && isServerCacheError(err) {
 		dcs.RESTMapper.Reset()
 
 		resourceList, err = dcs.DiscoveryClientCached.ServerResourcesForGroupVersion(gv.String())
+		if err != nil {
+			return nil, err
+		}
 	}
```
Comment (hausdorff) [inline pkg/clients/clients.go]: I'm confused about this conditional. Why can we ignore any error that isn't `isServerCacheError`?

File: pkg/clients/clients.go [hunk 4]
```diff
@@ -148,12 +148,15 @@ func (dcs *DynamicClientSet) ResourceClientForObject(obj *unstructured.Unstructu
...
-	if err != nil && len(resources) == 0 {
-		// Only return an error here if no resources were returned. Otherwise, process the partial list
-		// and return an error if no match is found.
-		return
+func (dcs *DynamicClientSet) gvkForKind(kind Kind) (*schema.GroupVersionKind, error) {
+	resources, err := dcs.DiscoveryClientCached.ServerPreferredResources()
+	if err != nil {
+		if discovery.IsGroupDiscoveryFailedError(err) {
+			// Ignore Group Discovery Failed errors. This type of error may cause
+			// later resource introspection to fail, but the error will be handled at that point.
```
Comment (hausdorff) [inline pkg/clients/clients.go]: For my own education: in this case, `resources` is empty, right? Also, `the error will be handled at that point` is referring to the return at the end of this method right? We might just reword it to say that instead -- not sure what it means as written.

File: pkg/clients/clients.go (hunk 3 above)
Comment (lblackstone) [reply pkg/clients/clients.go]: Fixed in subsequent commit

File: pkg/clients/clients.go (hunk 4 above)
Comment (lblackstone) [reply pkg/clients/clients.go]: I'll clarify the comment. `ServerPreferredResources` actually returns a partial list in case discovery failed for one or more groups. This ignores those errors and attempts to process the partial list. If the discovery failed for a relevant group, introspection will fail for the caller.
